xtra-libp2p-rollover/src/v_2_0_0/taker.rs

```diff
@@ -84,13 +85,42 @@ impl<E, O> Actor<E, O> {
                 v_2_0_0::PROTOCOL,
             ))
             .await
-            .context("Endpoint is disconnected")?
-            .context("No connection to peer")?
-            .await
-            .context("Failed to open substream")?;
+            .context("Endpoint is disconnected")?;
+
+        let substream = match substream.context("No connection to peer") {
+            Ok(substream) => substream,
+            Err(e) => {
+                self.emergency_drop_connection(peer_id, &e).await?;
+                return Err(e);
+            }
+        };
+
+        let substream = match substream.await.context("Failed to open substream") {
+            Ok(substream) => substream,
+            Err(e) => {
+                self.emergency_drop_connection(peer_id, &e).await?;
+                return Err(e);
+            }
+        };
```

Comment:
🤔 There are too many reasons why this could fail that do not warrant a disconnection:
- A failed negotiation because the parties disagree on the protocols will not be solved by this.
- The listener may take too long to respond to the message, but the connection between the two parties is still completely fine. We are just adding more work to the listener to handle an unnecessary reconnect.
- The dialer themself may take too long to perform the protocol! Perhaps the async work is not being scheduled in time because there's too much load on the application.

There are some `yamux::ConnectionError`s that may indicate that we should drop the connection, but not all:

```rust
pub enum ConnectionError {
    /// An underlying I/O error occured.
    Io(std::io::Error),
    /// Decoding a Yamux message frame failed.
    Decode(FrameDecodeError),
    /// The whole range of stream IDs has been used up.
    NoMoreStreamIds,
    /// An operation fails because the connection is closed.
    Closed,
    /// Too many streams are open, so no further ones can be opened at this time.
    TooManyStreams
}
```

In conclusion, I think one data point is insufficient to drop the connection. And giving this power to any one protocol does not seem wise either.

Having looked at the `yamux::ConnectionError`s above, I would be more interested in letting the `Endpoint` decide if it should drop a particular connection depending on the error (`Closed` and `Io` maybe?).

Reply:
> A failed negotiation because the parties disagree on the protocols will not be solved by this.

Yes, and it should not be. I would accept that a taker with an incompatible rollover protocol will reconnect or treat that specific error separate.

> The listener may take too long to respond to the message, but the connection between the two parties is still completely fine. We are just adding more work to the listener to handle an unnecessary reconnect.

I don't think this is a valid argument for what the user experience. The user failed to roll-over for 24h resulting in a force close. This would mean the listener was too busy to accept the substream for 24h.

> The dialer themself may take too long to perform the protocol! Perhaps the async work is not being scheduled in time because there's too much load on the application.

I don't think this is a valid argument for what the user experience. It is unlikely that the taker daemon was too busy to handle work within 24h. From our experience the taker daemon does not take many resources.

> Having looked at the yamux::ConnectionErrors above, I would be more interested in letting the Endpoint decide if it should drop a particular connection depending on the error (Closed and Io maybe?).

Thanks for having a look, I am happy to consider that. But I am not sure that would protect us from the endpoint not dropping a connection in some cases. The case where we keep failing to open a substream, but

Reply:
> > A failed negotiation because the parties disagree on the protocols will not be solved by this.
> 
> Yes, and it should not be. I would accept that a taker with an incompatible rollover protocol will reconnect or treat that specific error separate.

But with this code change you are mandating that every time they disagree on the protocol the taker reconnects, which doesn't solve anything.

> > The listener may take too long to respond to the message, but the connection between the two parties is still completely fine. We are just adding more work to the listener to handle an unnecessary reconnect.
> 
> I don't think this is a valid argument for what the user experience. The user failed to roll-over for 24h resulting in a force close. This would mean the listener was too busy to accept the substream for 24h.

The problem is we have no idea what actually happened, because we have no logs. We just have no actual proof that the problem is that the `taker` is not dropping the connection to the `maker` correctly. And even if it were the case, this would not be the right place to fix that.

As per my original message, if you drop the connection any time the `maker` is just lagging, you're making the situation worse for the lagging `maker`, who would have to handle the disconnect and the reconnect of this peer.

> > The dialer themself may take too long to perform the protocol! Perhaps the async work is not being scheduled in time because there's too much load on the application.
> 
> I don't think this is a valid argument for what the user experience. It is unlikely that the taker daemon was too busy to handle work within 24h. From our experience the taker daemon does not take many resources.

My original message is not trying to say that the user you're referencing experienced any of this. I'm just telling you that even if you were fixing that user's specific problem, you would be making matters worse in all situations where it doesn't apply.

> > Having looked at the yamux::ConnectionErrors above, I would be more interested in letting the Endpoint decide if it should drop a particular connection depending on the error (Closed and Io maybe?).
> 
> Thanks for having a look, I am happy to consider that. But I am not sure that would protect us from the endpoint not dropping a connection in some cases. The case where we keep failing to open a substream might not be covered.

_If_ we have a bug that causes takers to remain connected to a maker without a broken connection, I'm against fixing it in the rollover actor. That's what I'm advising.